core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/report/ReportEditViewModel.kt

```diff
@@ -0,0 +1,230 @@
+                reportTitle = currentReport.title,
+                reportOptions = Json.encodeToString(currentReport),
+            )
+            scheduleEntityCommitToSavedState(report, serializer = Report.serializer())
+            activeRepo.withDoorTransactionAsync {
+                try {
+                    if (entityUid == 0L) {
+                        activeRepo.reportDao().insert(report)
+                        println("Report options inserted successfully: ${report}")
+                    } else {
+                        activeRepo.reportDao().update(report)
+                        println("Report options updated successfully: ${report}")
+                    }
+                } catch (e: Exception) {
+                    println("Error updating report options: ${e.message}")
+                }
+            }
+        }
+    }
+
+    fun onEntityChanged(newOptions: ReportOptions2) {
+        _uiState.update { currentState ->
+            currentState.copy(reportOptions2 = newOptions)
+        }
+    }
+
+    fun onSeriesChanged(updatedSeries: ReportSeries2) {
+        _uiState.update { prev ->
+            prev.copy(
+                reportOptions2 = prev.reportOptions2.copy(
+                    series = prev.reportOptions2.series.replace(updatedSeries) {
+                        it.reportSeriesUid == updatedSeries.reportSeriesUid
+                    }
+                )
+            )
+        }
+    }
+
+    private fun onFilterChanged(filter2: ReportFilter3, seriesId: Int) {
+        if (filter2.reportFilterField != null) {
+            _uiState.update { prev ->
+                val updatedSeriesList = prev.reportOptions2.series.map { series ->
+                    if (series.reportSeriesUid == seriesId) {
+                        val updatedFilters =
+                            series.reportSeriesFilters?.toMutableList() ?: mutableListOf()
+                        updatedFilters.add(filter2)
+                        series.copy(reportSeriesFilters = updatedFilters)
+                    } else {
+                        series
+                    }
+                }
+
+                prev.copy(
+                    reportOptions2 = prev.reportOptions2.copy(
+                        series = updatedSeriesList
+                    )
+                )
+            }
+        }
+    }
+
+
+    fun onAddFilter(seriesId: Int) {
+        navigateForResult(
+            nextViewName = ReportFilterEditViewModel.DEST_NAME,
+            key = RESULT_KEY_REPORT,
+            currentValue = null,
+            serializer = Report.serializer(),
+            args = mapOf("reportSeriesUid" to seriesId.toString())
+        )
+    }
+
+
+    fun onAddSeries() {
+        _uiState.update { prev ->
+            prev.copy(
+                reportOptions2 = prev.reportOptions2.copy(
+                    series = prev.reportOptions2.series + ReportSeries2(
+                        reportSeriesUid = prev.reportOptions2.series.size + 1,
```

Comment:
This way of setting reportSeriesUid is flawed: if a user added series, then removed them, there is no gaurantee that you would not have duplicates.

You could use:

```
reportOptions2.series.maxOf { it.reportSeriesUid } + `
```